Source/SwiftLintFramework/Rules/CompilerProtocolInitRule.swift

```diff
@@ -0,0 +1,156 @@
+//
+//  CompilerProtocolInitRule.swift
+//  SwiftLint
+//
+//  Created by Marcelo Fabri on 12/31/16.
+//  Copyright © 2016 Realm. All rights reserved.
+//
+
+import Foundation
+import SourceKittenFramework
+
+public struct CompilerProtocolInitRule: ASTRule, ConfigurationProviderRule {
+    public var configuration = SeverityConfiguration(.warning)
+
+    public init() {}
+
+    public static let description = RuleDescription(
+        identifier: "compiler_protocol_init",
+        name: "Compiler Protocol Init",
+        description: "The initializers declared in compiler protocols such as `ExpressibleByArrayLiteral` " +
+                     "shouldn't be called directly.",
+        nonTriggeringExamples: [
+            "let set: Set<Int> = [1, 2]\n",
+            "let set = Set(array)\n"
+        ],
+        triggeringExamples: [
+            "let set = ↓Set(arrayLiteral: 1, 2)\n",
+            "let set = ↓Set.init(arrayLiteral: 1, 2)\n"
+        ]
+    )
+
+    public func validateFile(_ file: File, kind: SwiftExpressionKind,
+                             dictionary: [String : SourceKitRepresentable]) -> [StyleViolation] {
+        return violationRangesInFile(file, kind: kind, dictionary: dictionary).map {
+            StyleViolation(ruleDescription: type(of: self).description,
+                           severity: configuration.severity,
+                           location: Location(file: file, characterOffset: $0.location))
+        }
+    }
+
+    private func violationRangesInFile(_ file: File,
+                                       kind: SwiftExpressionKind,
+                                       dictionary: [String: SourceKitRepresentable]) -> [NSRange] {
+        guard kind == .call,
```

Comment:
The only one that would actually be computed in every loop would be `arguments` (since the other ones shouldn't really return `nil`). However, computing it earlier actually makes the rule a bit slower because most of times a `.call` won't be one of the forbidden ones, so we'd be computing `arguments` unnecessarily.